trinity/sync/full/chain.py

```diff
@@ -155,6 +144,87 @@ def __init__(self,
             len(headers), txs, timer.elapsed, head.block_number)
         return head.block_number
 
+    async def _download_receipts(self,
+                                 target_td: int,
+                                 all_headers: Tuple[BlockHeader, ...]) -> None:
+        """
+        Downloads and persists the receipts for the given set of block headers.
+        Receipts are requested from all peers in equal sized batches.
+        """
+        # Post-Byzantium blocks may have identical receipt roots (e.g. when they have the same
+        # number of transactions and all succeed/failed: ropsten blocks 2503212 and 2503284),
+        # so we do this to avoid requesting the same receipts multiple times.
+        headers = tuple(unique(
+            (header for header in all_headers if not _is_receipts_empty(header)),
+            key=operator.attrgetter('receipt_root'),
+        ))
+
+        while headers:
+            # split the remaining headers into equal sized batches for each peer.
+            peers = cast(Tuple[ETHPeer, ...], self.peer_pool.get_peers(target_td))
+            batch_size = math.ceil(len(headers) / len(peers))
+            batches = tuple(partition_all(batch_size, headers))
+
+            # issue requests to all of the peers and wait for all of them to respond.
+            requests = tuple(
+                self._get_receipts(peer, batch)
+                for peer, batch
+                in zip(peers, batches)
+            )
+            responses = await self.wait(asyncio.gather(
+                *requests,
+                loop=self.get_event_loop(),
```

Comment:
I guess if we start seeing this a lot, we can add a `BaseService.gather()` to turn this into:

```py
responses = await self.gather(*requests)
```